extensions/amp-ad-network-doubleclick-impl/0.1/doubleclick-a4a-config.js

```diff
@@ -77,55 +83,115 @@ export const BETA_ATTRIBUTE = 'data-use-beta-a4a-implementation';
+   * @param {!Window} win
+   * @return {boolean}
+   */
+  isCdnProxy(win) {
+    const googleCdnProxyRegex =
+      /^https:\/\/([a-zA-Z0-9_-]+\.)?cdn\.ampproject\.org/;
+    return !!googleCdnProxyRegex.test(win.location.origin);
   }
-  // See if in holdback control/experiment.
-  let experimentId;
-  const urlExperimentId = extractUrlExperimentId(win, element);
-  if (urlExperimentId != undefined) {
-    experimentId = URL_EXPERIMENT_MAPPING[urlExperimentId];
-    dev().info(
-        TAG, `url experiment selection ${urlExperimentId}: ${experimentId}.`);
-  } else {
-    // Not set via url so randomly set.
+
+  /** Whether Fast Fetch is enabled
+   * @param {!Window} win
+   * @param {!Element} element
+   * @return {boolean}
+   */
+  isA4aEnabled(win, element) {
+    let experimentId;
+    if ('useSameDomainRenderingUntilDeprecated' in element.dataset ||
+        element.hasAttribute('useSameDomainRenderingUntilDeprecated') ||
+        !this.supportsCrypto(win)) {
+      return false;
+    }
+    let experimentName = DFP_CANONICAL_FF_EXPERIMENT_NAME;
+    if (!this.isCdnProxy(win)) {
+      experimentId = this.selectExperiment(win, element, [
+        DOUBLECLICK_EXPERIMENT_FEATURE.CANONICAL_CONTROL,
+        DOUBLECLICK_EXPERIMENT_FEATURE.CANONICAL_EXPERIMENT,
+      ], DFP_CANONICAL_FF_EXPERIMENT_NAME);
+    } else {
+      experimentName = DOUBLECLICK_A4A_EXPERIMENT_NAME;
+      if (element.hasAttribute(BETA_ATTRIBUTE)) {
+        addExperimentIdToElement(BETA_EXPERIMENT_ID, element);
+        dev().info(TAG, `beta forced a4a selection ${element}`);
+        return true;
+      }
+      // See if in holdback control/experiment.
+      const urlExperimentId = extractUrlExperimentId(win, element);
+      if (urlExperimentId != undefined) {
+        experimentId = URL_EXPERIMENT_MAPPING[urlExperimentId];
+        dev().info(
+            TAG,
+            `url experiment selection ${urlExperimentId}: ${experimentId}.`);
+      } else {
+        experimentId = this.selectExperiment(win, element, [
+          DOUBLECLICK_EXPERIMENT_FEATURE.HOLDBACK_INTERNAL_CONTROL,
+          DOUBLECLICK_EXPERIMENT_FEATURE.HOLDBACK_INTERNAL],
+            DOUBLECLICK_A4A_EXPERIMENT_NAME);
+      }
+    }
+    if (experimentId) {
+      addExperimentIdToElement(experimentId, element);
+      forceExperimentBranch(win, DOUBLECLICK_A4A_EXPERIMENT_NAME, experimentId);
+    }
+    return ![DOUBLECLICK_EXPERIMENT_FEATURE.HOLDBACK_EXTERNAL,
+      DOUBLECLICK_EXPERIMENT_FEATURE.HOLDBACK_INTERNAL,
+      DOUBLECLICK_EXPERIMENT_FEATURE.SFG_CONTROL_ID,
+      DOUBLECLICK_EXPERIMENT_FEATURE.SFG_EXP_ID,
+      DOUBLECLICK_EXPERIMENT_FEATURE.CANONICAL_CONTROL,
+    ].includes(experimentId);
+  }
+
+  /**
+   * @param {!Window} win
+   * @param {!Element} element
+   * @param {!Array} selectionBranches
+   * @param {!string} experimentName}
+   * @return {?string} Experiment branch ID.
+   */
+  selectExperiment(win, element, selectionBranches, experimentName) {
```

Comment:
Any reason you can't make this private?  Consider renaming to maybeSelectExperiment given that sometimes it does not depending on experiment fraction